usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -2917,6 +2922,175 @@ CK_RV ep11tok_aes_xts_pkey_check(STDLL_TokData_t *tokdata, CK_ATTRIBUTE_PTR attr
+ * unwrap it by the wrap key
+ */
+static CK_RV import_aes_xts_key(STDLL_TokData_t *tokdata, SESSION *sess,
+                                OBJECT *aes_xts_key_obj,
+                                CK_BYTE *blob, size_t *blob_size)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_BYTE cipher[MAX_BLOBSIZE];
+    CK_ULONG clen = sizeof(cipher);
+    CK_BYTE csum[MAX_CSUMSIZE];
+    size_t cslen = sizeof(csum);
+    CK_BYTE iv[AES_BLOCK_SIZE];
+    size_t blob_size2 = *blob_size;
+    CK_MECHANISM mech = { CKM_AES_CBC_PAD, iv, AES_BLOCK_SIZE };
+    CK_RV rc;
+    CK_ATTRIBUTE_PTR p_attrs = NULL;
+    CK_ULONG attrs_len = 0;
+    CK_ATTRIBUTE_PTR new_p_attrs = NULL;
+    CK_ULONG new_attrs_len = 0;
+    CK_ATTRIBUTE *chk_attr = NULL, *attr = NULL;
+    unsigned char *ep11_pin_blob = NULL;
+    CK_ULONG ep11_pin_blob_len = 0;
+    ep11_session_t *ep11_session = (ep11_session_t*) sess->private_data;
+
+    rc = template_attribute_get_non_empty(aes_xts_key_obj->template, CKA_VALUE,
+                                          &attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Could not find CKA_VALUE for the key.\n");
+        return rc;
+    }
+
+    /* tell ep11 the attributes the user specified */
+    rc = build_ep11_attrs(tokdata, aes_xts_key_obj->template, &p_attrs,
+                          &attrs_len, CKK_AES, CKO_SECRET_KEY, -1, &mech);
+    if (rc != CKR_OK)
+        goto import_aes_xts_key_end;
+
+    rc = ep11tok_aes_xts_pkey_check(tokdata, p_attrs, attrs_len);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s EP11 AES XTS is not supported: rc=0x%lx\n",
+                    __func__, rc);
+        goto import_aes_xts_key_end;
+    }
+
+    memset(cipher, 0, sizeof(cipher));
+    memcpy(iv, "1234567812345678", AES_BLOCK_SIZE);
+
+    /*
+     * calls the ep11 lib (which in turns sends the request to the card),
+     * all m_ function are ep11 functions
+     */
+    RETRY_START(rc, tokdata)
+        rc = dll_m_EncryptSingle(ep11_data->raw2key_wrap_blob,
+                                 ep11_data->raw2key_wrap_blob_l, &mech,
+                                 attr->pValue, attr->ulValueLen / 2,
+                                 cipher, &clen, target_info->target);
+    RETRY_END(rc, tokdata, sess)
+
+    if (rc != CKR_OK) {
+        rc = ep11_error_to_pkcs11_error(rc, sess);
+        TRACE_ERROR("%s encrypt ksize=0x%lx clen=0x%lx rc=0x%lx\n", __func__,
+                    attr->ulValueLen / 2, clen, rc);
+        goto import_aes_xts_key_end;
+    } else {
+        TRACE_INFO("%s encrypt ksize=0x%lx clen=0x%lx rc=0x%lx\n", __func__,
+                   attr->ulValueLen / 2, clen, rc);
+    }
+
+    rc = check_key_attributes(tokdata, CKK_AES, CKO_SECRET_KEY, p_attrs,
+                              attrs_len, &new_p_attrs, &new_attrs_len, -1);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s AES XTS check private key attributes failed with "
+                    "rc=0x%lx\n", __func__, rc);
+        return rc;
```

Comment:
This should be `goto import_aes_xts_key_end;` instead of the return.
This was already wrong in `rawkey_2_blob()` where you copied it from ;-)

If you like, add an additional commit to this PR fixing this in `rawkey_2_blob()`.
The commit title would be for example `EP11: Fix potential memory leak when importing symmetric keys`.